File: assembly/actions/transfer.ts
import { ActionSchema, HandlerResultSchema, StateSchema } from '../schemas';
import { Transaction } from '../imports/smartweave/transaction';
import { console } from '../imports/console';

export function transfer(
  state: StateSchema,
  action: ActionSchema
): HandlerResultSchema {
  const target = action.target;
  const qty = action.qty;
  const caller = Transaction.owner();

  if (qty <= 0 || caller == target) {
    throw new Error('[CE:ITT] Invalid token transfer');
  }

  if (!state.balances.has(caller) || state.balances.get(caller) < qty) {
    throw new Error(
      `[CE:NEB] Caller balance not high enough to send ${qty} token(s)!`
    );
  }

  // Lower the token balance of the caller
  state.balances.set(caller, state.balances.get(caller) - qty);
  if (!state.balances.has(target)) {
    state.balances.set(target, qty);
  } else {
    state.balances.set(target, state.balances.get(target) + qty);
  }

  console.log(`Balance ${state.balances.get(target)}`);
  console.log(`Balance ${state.balances.get(caller)}`);

  return {
    state,
    result: null,
  };
}

File: rust/src/actions/transfer.rs
use crate::error::ContractError::{CallerBalanceNotEnough, TransferAmountMustBeHigherThanZero};
use crate::state::State;
use crate::action::ActionResult;
use crate::contract_utils::handler_result::HandlerResult;
use crate::contract_utils::js_imports::{log, SmartWeave, Transaction};

pub fn transfer(mut state: State, qty: u64, target: String) -> ActionResult {
    log(("caller ".to_owned() + &SmartWeave::caller()).as_str());
    log(("Transaction owner ".to_owned() + &Transaction::owner()).as_str());

    if qty == 0 {
        return Err(TransferAmountMustBeHigherThanZero);
    }

    let caller = Transaction::owner();
    let balances = &mut state.balances;

    // Checking if caller has enough funds
    let caller_balance = *balances.get(&caller).unwrap_or(&0);
    if caller_balance < qty {
        return Err(CallerBalanceNotEnough(caller_balance));
    }

    // Update caller balance
    balances.insert(caller, caller_balance - qty);

    // Update target balance
    let target_balance = *balances.get(&target).unwrap_or(&0);
    balances.insert(target, target_balance + qty);

    Ok(HandlerResult::NewState(state))
}
